### PR TITLE
OSHMEM/ATOMICS: added C implementation of and/or/xor ops

### PR DESCRIPTION
**Pull request by hoopoepg**
- added implementation and/or/xor operations for post and
  fetch-op notations
- implemented basic and UCX transports, mxm added
  NON-IMPLEMENTED wrapper
- updated C interfaces only (fortran will be added later)
- existing API is not updated to spec v1.4

Signed-off-by: Sergey Oblomov <sergeyo@mellanox.com>

### CODE REVIEW

**Comment by ibm-ompi** [comment]
The IBM CI (GNU Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/00e377d6d3e9eadd0335fafa54332edb

**Comment by ibm-ompi** [comment]
The IBM CI (XL Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/3b271624c46b11d23941c8e99caf072e

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/9e930eecace3705c840ee5049bdd4dfe

**Comment by hoopoepg** [comment]
bot:mlx:retest

**Comment by hoopoepg** [comment]
bot:retest

**Comment by hoopoepg** [comment]
@brminich could you review?

File: oshmem/mca/atomic/atomic.h [hunk 1]
```diff
@@ -35,6 +35,49 @@ BEGIN_C_DECLS
...
+                                                                           \
+        size = sizeof(value);                                              \
+        rc = MCA_ATOMIC_CALL(op(                                           \
+            (void*)target,                                                 \
+            value,                                                         \
+            size,                                                          \
+            pe));                                                          \
+        RUNTIME_CHECK_RC(rc);                                              \
+                                                                           \
+        return ;                                                           \
```
Comment (brminich) [inline oshmem/mca/atomic/atomic.h]: minor: extra whitespace after `return`

File: oshmem/shmem/c/shmem_and.c [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+#include "oshmem/include/shmem.h"
+#include "oshmem/include/shmemx.h"
+
+#include "oshmem/runtime/runtime.h"
+#include "oshmem/op/op.h"
+#include "oshmem/mca/atomic/atomic.h"
+
+/*
+ * These routines perform an atomic 'and' operation.
+ * The atomic 'and' routines cwoperates 'and' value to the data at address target on PE pe.
```
Comment (brminich) [inline oshmem/shmem/c/shmem_and.c]: `cwoperates`?

File: oshmem/shmem/c/shmem_for.c [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+#include "oshmem/runtime/runtime.h"
+
+#include "oshmem/op/op.h"
+#include "oshmem/mca/atomic/atomic.h"
+
+/*
+ * These routines perform an atomic fetch-and-or operation.
+ * The fetch and or routines retrieve the value at address target on PE pe, and update
+ * target with the result of 'or' operation value to the retrieved value. The operation
+ * must be completed * without the possibility of another process updating target between
```
Comment (brminich) [inline oshmem/shmem/c/shmem_for.c]: extra `*`

File: oshmem/shmem/c/shmem_fxor.c [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#include "oshmem/runtime/runtime.h"
+
+#include "oshmem/op/op.h"
+#include "oshmem/mca/atomic/atomic.h"
+
+/*
+ * These routines perform an atomic fetch-and-xor operation.
+ * The fetch and xor routines retrieve the value at address target on PE pe, and update
+ * target with the result of 'xor' operation value to the retrieved value. The operation
+ * must be completed * without the possibility of another process updating target between
```
Comment (brminich) [inline oshmem/shmem/c/shmem_fxor.c]: extra `*`

File: oshmem/shmem/c/shmem_or.c [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+#include "oshmem/include/shmem.h"
+#include "oshmem/include/shmemx.h"
+
+#include "oshmem/runtime/runtime.h"
+#include "oshmem/op/op.h"
+#include "oshmem/mca/atomic/atomic.h"
+
+/*
+ * These routines perform an atomic 'or' operation.
+ * The atomic 'or' routines cwoperates 'or' value to the data at address target on PE pe.
```
Comment (brminich) [inline oshmem/shmem/c/shmem_or.c]: `cwoperates`?

File: oshmem/shmem/c/shmem_xor.c [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+#include "oshmem/include/shmem.h"
+#include "oshmem/include/shmemx.h"
+
+#include "oshmem/runtime/runtime.h"
+#include "oshmem/op/op.h"
+#include "oshmem/mca/atomic/atomic.h"
+
+/*
+ * These routines perform an atomic 'xor' operation.
+ * The atomic 'xor' routines cwoperates 'xor' value to the data at address target on PE pe.
```
Comment (brminich) [inline oshmem/shmem/c/shmem_xor.c]: `cwoperates`?

File: oshmem/mca/atomic/atomic.h (hunk 1 above)
Comment (hoopoepg) [reply oshmem/mca/atomic/atomic.h]: removed

File: oshmem/shmem/c/shmem_and.c (hunk 2 above)
Comment (hoopoepg) [reply oshmem/shmem/c/shmem_and.c]: fixed

File: oshmem/shmem/c/shmem_for.c (hunk 3 above)
Comment (hoopoepg) [reply oshmem/shmem/c/shmem_for.c]: removed

File: oshmem/shmem/c/shmem_fxor.c (hunk 4 above)
Comment (hoopoepg) [reply oshmem/shmem/c/shmem_fxor.c]: removed

File: oshmem/shmem/c/shmem_or.c (hunk 5 above)
Comment (hoopoepg) [reply oshmem/shmem/c/shmem_or.c]: fixed

File: oshmem/shmem/c/shmem_xor.c (hunk 6 above)
Comment (hoopoepg) [reply oshmem/shmem/c/shmem_xor.c]: fixed

**Comment by hoopoepg** [comment]
bot:retest

**Comment by hoopoepg** [comment]
bot:retest

**Comment by hoopoepg** [comment]
bot:retest

**Comment by hoopoepg** [comment]
@jladd-mlnx 👍

**Pull request merged**